### ISSUE TITLE
custom (de)serialization of Date when snapshotting

### ISSUE THREAD

**Issue by faassen** [mobxjs/mobx-state-tree#662]
Date fields are the only types in mobx-state-tree that cannot be unambiguously converted to and from JSON. You chose to use Unix timestamps (as those are apparently in use by a lot of backends), but I'm dealing with a backend that generates ISO 8601 strings instead. I think we can acknowledge that backends handle dates in range of different ways, and that a string representation is not uncommon at all. json-schema for instance also treats date-time as a specific format for strings (a restricted profile of ISO 8601, defined by rfc3339, section 5.6).

I could either modify the backend to work with timestamps, or create lifecycle hooks to convert ISO 8601 strings into numbers and vice versa, but both approaches are inconvenient; for the backend I'd have to deviate from the normal DateTime API, and for the frontend I'd need to remember to write custom hooks for each model that uses a Date field. There's doesn't appear to be enough information to easily automate such a hook: preProcessSnapshot doesn't have information about the type available, though I haven't explored middleware yet.

It would be much more convenient if there were a centralized policy where I could determine how Dates are converted to and from JSON. I realize centralized policies are something to avoid if possible, and it would make snapshots generated through different policies incompatible with each other, but it would seem to be the simplest way to help integrate mobx-state-tree with a backends with a different interpretation of date/time.

**Comment by robinfehr** [comment]
efforts for custom primitive types are already ongoing have a look at #609  -
as for now, couldn't you make you own Date Model and reuse this whenever you're in need of a date? that way you'd not have to add the hook everywhere.
with e.g. one property  `date: types.Date()` and use postProcessSnapshot

```js
const DateModel = types.model('DateModel', {
    date: types.Date
}).actions(self => ({
    postProcessSnapshot(snapshot) {
       return {
           date: convert(snapshot.date)
       }
    }
})

**Comment by faassen** [comment]
I think this is slightly different than creating a custom primitive type: I want to modify the behavior of an existing primitive type. I don't want to have to forbid the developers form using types.Date and use my own new primitive type instead, I want to modify the way mobx-state-tree serializes types.Date. 

Custom primitives do look useful, and Decimals are one use case where they'd be very useful indeed. Pretty soon I'm going to need a decimal type that serializes to strings.

I don't think your DateModel suggestion will work as won't that create a subobject with a "date" entry everywhere? That still won't integrate smoothly with a backend that expect date strings -- it will make integration with this backend harder, not easier.

(for now I used the preprocess hook to work around this)

**Comment by robinfehr** [comment]
i see your backed is already static in terms of:  the api is already given.
in such case my suggestion does not really help yes.

about type information - soon there will be a `reflect(node)` which might help you.

e.g. 

```
// within an action
reflect(self).properties[propertyName].describe() // will give you back the type
```

**Comment by mweststrate** [comment]
I think at this moment you can already achieve it by using custom references to implement a custom primitive (if not using typescript): https://github.com/mobxjs/mobx-state-tree#customizable-references

**Comment by faassen** [comment]
It's nice that I can create a custom primitive with custom references. I take it the "identifier" in this API is the value in the JSON serialization? I imagine this logic happens before the serialization hooks.

As I tried to argue, there is no canonical JSON serialization of dates. Some people use timestamps, some people use strings. mobx-state-tree currently uses timestamps. It makes sense to me that people should be able to write models that express that they are dealing with dates (using the types.Date type) without committing to a particular serialization approach.

So this is a question of the philosophy mobx-state-tree wants to express: is it part of the definition of a primitive how it is expressed as JSON, or are primitives only about  specifying models?

So perhaps the philosophy of mobx-state-tree is that primitives also express details of JSON serialization. I'll note that's not quite the case already: I can modify the serialization procedure by implementing a few hooks. So they currently express the default serialization (and the expected arguments and return values of said hooks). I'm effectively arguing for another, more global hook, where I can influence the serialization of particular types.

I expect decimals have a similar issue -- are they to be serialized as strings (which can fully express decimals -- Python backend tend this way) or as JS numbers (JSON-Schema does this)? (but it gets more involved as you can also debate their representation in JS; it make sense to me to use different primitives in that case)

**Comment by robinfehr** [comment]
@faassen, @mweststrate  in my opinion
types.Date shouldn't be called a primitive type in MST since javascript doesn't declare it as such either. leads to confusion since types.Date becomes a number on getSnapshot.

@faassen 
as for the philosophy/ sense: 
the mst types are used for the serialization / snapshot generation.
a type describes how it's serialized version looks like. 
(in each type there is a `getSnapshot()` method implemented)

therefore - if you build a custom type on top of a primitive type such should tell, how the serialized version looks like. (e.g. decimal will become a string)
 
as for the hooks: 
they're opening the door to many use cases. don't think they're to blame 🗡 

btw. we haven't mentioned the approach yesterday of using a view. (just for illustration purposes)
https://codesandbox.io/s/8k4j6j3xo0

the approach using the custom primitive type would solve that a bit more elegant though since there would be no confusion about which property should be used.
https://codesandbox.io/s/x7mxr0kr4w

**Comment by faassen** [comment]
Concerning philosophy: I realize that mst types are used to drive serialization too, but they're also used for other purposes -- promise to a developer that a model field is a number, say. 

I'm not suggesting types should stop driving serialization, I'm just suggesting it might be useful to influence the process of (de)serialization in a primitive type driven way (just like we can already influence model (de)serialization).

There are other use cases here too: what if you want to support multiple backends with the same codebase, and one backend serializes datetime in one way, the other in another?

**Comment by robinfehr** [comment]
hm you mean something like an overwrite-able `getSnapshot()` of a type in general right? in which you could decide upon the backend-type/serialization.
(e.g. not using `node.storedValue.getTime()` but another format) 

I'm concerned that would mean exposing too much - if you have a look at the getSnapshot of the `model.ts` for example.

I like the idea 👍 but couldn't that be achieved using a simple `if` within a custom ref type :) e.g.

```js
const CustomDate = types.reference(types.string, {
    set(value) { // i think the setter does currently not work with primitive types :)
      if (currentBackend === 'backend1')
          return new Date(value).getTime()
      return new Date(value).toString()
    }
  })
```
combined with an union I guess it would also be achieveable to have a new Date(value).getTime() and such mixed.

**Comment by faassen** [comment]
Yes, a way to override how a primitive type is "snapshotted" similar to the way you can override how a model is snapshotted. But for primitive types it'll have to be a bit more global.

I guess one way to do is to create a custom "types", override the behavior in it, and use that everywhere. But I'm not sure that is enough for the use case where you want to support multiple backend formats in the same application. But of course that use case may not be so important.

**Comment by faassen** [comment]
@robinfehr could you explain a bit of the reasoning about why this issue is closed? Was it decided you don't want different ways to export built-in types and I should instead go for a custom type? Is a fix along the way as part of another issue? I'm confused.

**Comment by mweststrate** [comment]
@faassen If I understand the question correctly, this is basically already planned: https://github.com/mobxjs/mobx-state-tree/pull/609/files 

I noticed that there was no actual issue for, so created one, but the feature is WIP and I hope to finish it in the upcoming weeks: https://github.com/mobxjs/mobx-state-tree/issues/673

**Comment by robinfehr** [comment]
@faassen urgh sorry for closing eagerly - thought it was answered satisfyingly mentioning the ongoing work around the custom types as the current way to go / wait for.

as for a even more general solution as you'd want to have it - I was actually going to try to pipe the postProcessSnapshot and other hooks - even if not declared explicitly as a member - to the middleware so one could use the reflection api and add a middleware that intercepts the postProcessSnapshot to do such things. can't yet promise anything - it might not be a good idea, but I'll see about that soon 🔢 .

as for the exposing - have a quick look at the different implementations of the getSnapshot for each type, i don't think that should be exposed. maybe you'd think differently and we could discuss further. (a community is stronger than one! 💃 )

**Comment by faassen** [comment]
While custom primitives could be used, there are some potential issues with this.

* I argue that if you want people to use a Date, they should use types.Date, and not have to introduce a new type just because their serialization requirements are different. I would prefer to avoid having to introduce a new Date type for my project -- I need to communicate this separately from the official docs and it would behave exactly like Date from the perspective of the JS API anyway.

* A more complex use case I'm not sure you want to support (yet): consider a web application that needs to talk to two different backends, but each use a different way to serialize dates. How would I support this? I don't want to write two models. I don't think I can even introduce a custom Date type that gets parameterized with the backend type, right?

My goal in all this is to argue for this use case: affect the serialization of primitive types without having to write custom snapshotting logic for each model. This use case comes up in the case of Dates, and also for Decimals (which could have a number or string serialization). 

If the use case is deemed something you don't want to support, as primitive types really have only 1 form of serialization completely determined by the type (unlike model types), then we can close this issue. If so, I would suggest you add some text that explains this in the documentation, perhaps as part of #673.

**Comment by robinfehr** [comment]
as for the first:
I agree that types.Date can be misleading - when we'll have custom types..I personally would vote to remove it.

as for the second:
you could use conditions e.g.`if, swtich etc.` within the `getter`/ `setter` to have a conditional behavior

**Comment by mweststrate** [comment]
@faassen If I understand you correctly, this is exactly what #673 will solve (please tell us if we miss something!):

```javascript
`const MyBackendDate = types.custom<S, T>({
   get(value: S): T {
        const parsedDate = someWeirdParsingLogicFromTheString(value)
        return parseDate() // a normal JS Date object
    },
    set(value: T): S {
        const serializedForm = someWeirdUnparseLogicFromTheDate(value)
        return serializedForm // e.g. "Tue 7 Octobr '18'"
    }
})

const Appointment = types.model("appointment", {
   subject: types.string,
  date: MyBackendDate
})

const appointment = Appointment.create({
  subject: "Buy snacks",
  date:  "Tue 7 Octobr '18'"
})

console.log(appointment.date instanceof Date) // true
```

**Comment by faassen** [comment]
If I understand your example correctly you need to use MyBackendDate instead of types.Date if you want to support another backend.

I don't want to be a bother, but it's not *exactly* what I mean. This approach forces everyone who has a backend with string dates to actively avoid the use of types.Date. Instead they have to use MyBackendDate. That's fine if I develop by myself; I can remember. 

But in a team of developers you can't tell someone to just read the mobx-state-tree documentation anymore. Instead you need to tell them to read the documentation and also never to use types.Date but to use MyBackendDate instead.

That's not a big deal (project-specific usage rules happen with every framework) and I can live with this, but I wanted to make my point clear: it would be clearer if I could map MyBackendDate onto types.Date somehow.

Alternatively mobx-state-tree could unprivilege types.Date and tell everyone to implement their own Date and types.Date is just a convenience for a particular type of backend. But that seems harder to explain.

To put it another way: mobx-state-tree is quite similar to an ORM in some ways. ORMs abstract over database-specific details; perhaps one database uses a timestamp for dates and another uses a custom date type, but when I use the ORM I can forget this. Here I want to be able to abstract over serialization specific details.

Does that make sense?

**Comment by stoefln** [comment]
@faassen totally agree. I feel like types.Date should have the option to be configurable to read and write a certain forumat during de/serialization.
